core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -255,17 +273,162 @@ private void rebuildIndex() throws Exception {
+      TransactionConfiguration txConfig = config.transaction();
+      PrivateMetadata.Builder builder = new PrivateMetadata.Builder();
+      if (txConfig.transactionMode() == TransactionMode.TRANSACTIONAL && txConfig.lockingMode() == LockingMode.OPTIMISTIC) {
+         builder.entryVersion(new SimpleClusteredVersion(1, 1));
+      }
+
+      // Async XSite
+      if (config.sites().hasAsyncEnabledBackups()) {
+         String siteName = cache.getCacheManager().getTransport().localSiteName();
+         IracEntryVersion version = new IracEntryVersion(Collections.singletonMap(siteName, TopologyIracVersion.newVersion(1)));
+         builder.iracMetadata(new IracMetadata(siteName, version));
+      }
+      return builder.build();
+   }
+
+   private void migrateFromV12_0()  throws Exception {
+      // ISPN-13128 Corrupt migration data can only be created with default marshaller
+      if (ctx.getGlobalConfiguration().serialization().marshaller() == null) {
+         migrateCorruptDataV12_0();
+         return;
+      }
+
+      // Data is not corrupt, so simply update file magic
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startMigratingPersistenceData(cacheName);
+      try {
+         channel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+      } catch (IOException e) {
+         throw PERSISTENCE.persistedDataMigrationFailed(cacheName, e);
+      }
+      rebuildIndex();
+   }
+
+   private void migrateCorruptDataV12_0() {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      // Day before the release of Infinispan 10.0.0.Final
+      // This was the first release that SFS migrations on startup could be migrated from via ISPN 10 -> 11 -> 12
+      long sanityEpoch = LocalDate.of(2019, 10, 26)
+            .atStartOfDay()
+            .atZone(ZoneId.systemDefault())
+            .toInstant()
+            .toEpochMilli();
+      long currentTs = timeService.wallClockTime();
+
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_12_0), 0);
+
+         long oldFilePos = MAGIC_12_0.length;
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            // EOF reached
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            // Semantic check to find valid FileEntry
+            // fe.keyLen = 0 is valid, however it means we should skip the entry
+            if (fe.size <= 0 || fe.keyLen <= 0 || fe.dataLen <= 0 || fe.metadataLen < 0 || fe.internalMetadataLen < 0 || fe.expiryTime < -1) {
```

Comment:
Can we add something like `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` as well? Just to eliminate waiting for a throwable to determine it was corrupted

Reply:
isn't `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` always true for corrupted entries?

Reply:
Yeah, which is why we want to skip that byte :)

Reply:
> isn't `fe.size < fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen` always true for corrupted entries?

I don't think so if the corrupt entry was subsqeuently "freed" and `fe.keyLen = 0`.

Reply:
but in that case, you don't want to skip. you want to convert to the correct format

Reply:
But I guess in that case `fe.keyLen <= 0` protects against that